expand.py

```diff
@@ -30,7 +30,7 @@
 inc     = re.compile("^\$\$INCLUDE\((.*)\)")
 idxre   = re.compile("^INDEX([0-9]*)")
 doubledollar = re.compile("^(.*?)\$\$")
-keyword      = re.compile("^(ROOT|SUBSTR|UP|LOOP|IF|INCLUDE|TRANSLATE|COUNT|NAME)\(|^(CALC|IFCALC)\{")
+keyword      = re.compile("^(ROOT|SUBSTR|UP|LOOP|IF|INCLUDE|TRANSLATE|COUNT|NAME)\(|^(ASSIGN|CALC|IFCALC)\{")
```

Comment:
On that tangent @ZLLentz, are there any ongoing efforts to upgrade `expand.py2` to `expand.py3`? I have a local copy that I was working with to better comprehend the script and shoved it through `2to3` as a first pass. I have some ideas for simplifying some of the regex for parsing purposes and want to test it more in my local dev area before I claim any sort of improvement (_I anticipate I'm breaking at least one thing_)

Reply:
There haven't been any ongoing efforts to update this repo to py3. There isn't a huge amount of short-term benefit for such an effort unless we're pairing it with some other capability expansions.

The default python os install on rhel7 is python2, so when you `make` we have some guarantee that the interpreter is available.

Perhaps the story changes when we move to rocky9, which has only python3 installed, and perhaps now with the conda setups we could update it without too much hassle. In fact, with rocky9, it may become important to update to py3...

Reply:
> Unrelated to my review: maybe a future version of the IOC templating could leverage standard tools like jinja instead of using custom $$ macros. Jinja has essentially any option we'd ever want when it comes to assigning variables, doing loops, if statements, calculations, etc.: https://jinja.palletsprojects.com/en/3.1.x/templates/

Yeah, reinventing the wheel 'r' us... :-(

Reply:
> Perhaps the story changes when we move to rocky9, which has only python3 installed, and perhaps now with the conda setups we could update it without too much hassle. In fact, with rocky9, it may become important to update to py3...

Since this is an inevitable future and we only have Mike for a few more weeks, it might be worthwhile to start this sooner vs. later. I can volunteer to make a branch on this and track my changes/efforts/questions until someone more appropriate takes over/has time. Some things like this $$ASSIGN macro and other nuances are very relevant for some changes I want to make to SmarAct IOCs (and maybe integrating into jinja templates, depending on how we want to handle mapping values from additional scripts to set some default PV behavior for controls).

Reply:
> it might be worthwhile to start this sooner vs. later

I think you're right about this